### PR TITLE
Offline Support - Offline Actions Queuing + Running

### PR DESCRIPTION
**Pull request by polqf**
### 🔗 Issue Link
Part of https://stream-io.atlassian.net/browse/CIS-1484

### 🎯 Goal

This PR adds the final bits to get the offline queued requests executed when the connection is recovered

### 🛠 Implementation

`OfflineRequestsRepository` handles both the enqueuing and the execution of the requests when needed.
When running the queued requests, it basically passes the requests on to the APIClient, and waits for its result.

- If the requests succeeds -> The request is removed from the pending ones
- If the request fails with a connection error -> The request is kept to be executed once the connection is back (we are not putting it back at the queue to make sure we respect the order)
- If the request fails with any other error -> We are dismissing the request, and removing it from the queue

### ☑️ Checklist

- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] This change follows zero ⚠️ policy (required)
- [ ] Changelog is updated with client-facing changes
- [ ] New code is covered by unit tests
- [ ] Affected documentation updated (docusaurus, tutorial, CMS (task created)

### CODE REVIEW

File: Sources/StreamChat/APIClient/APIClient.swift [hunk 1]
```diff
@@ -94,9 +99,8 @@ class APIClient {
         endpoint: Endpoint<Response>,
         completion: @escaping (Result<Response, Error>) -> Void
     ) {
-        guard isRecoveryModeOn else {
+        if !isRecoveryModeOn {
             log.assertionFailure("We should not call this method if not in recovery mode")
```
Comment (adamrushy) [inline Sources/StreamChat/APIClient/APIClient.swift:103]: Before I forget, we should probably assert here or at least return out?

File: Sources/StreamChat/APIClient/APIClient.swift [hunk 2]
```diff
@@ -252,12 +270,14 @@ class APIClient {
...
+        log.debug("Entering recovery mode", subsystems: .offlineSupport)
         isRecoveryModeOn = true
         operationQueue.isSuspended = true
     }
 
     func exitRecoveryMode() {
         // Once recovery is done, regular requests can go through again.
+        log.debug("Leaving recovery mode", subsystems: .offlineSupport)
         isRecoveryModeOn = false
         operationQueue.isSuspended = false
```
Comment (adamrushy) [inline Sources/StreamChat/APIClient/APIClient.swift:282]: Before I forget, does this automatically kick off the OQ? From memory, it doesn't but I could be wrong...

Comment (polqf) [reply Sources/StreamChat/APIClient/APIClient.swift:282]: It is resuming the Operation Queue just by setting this value 😄

Comment (polqf) [reply Sources/StreamChat/APIClient/APIClient.swift:282]: `You may continue to add operations to a queue that is suspended but those operations are not scheduled for execution until you change this property to false.`

Comment (adamrushy) [reply Sources/StreamChat/APIClient/APIClient.swift:282]: No problem it was just one to check 💯

File: Sources/StreamChat/Repositories/OfflineRequestsRepository.swift [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+    init(database: DatabaseContainer, apiClient: APIClient) {
+        self.database = database
+        self.apiClient = apiClient
+    }
+
+    /// - If the requests succeeds -> The request is removed from the pending ones
+    /// - If the request fails with a connection error -> The request is kept to be executed once the connection is back (we are not putting it back at the queue to make sure we respect the order)
+    /// - If the request fails with any other error -> We are dismissing the request, and removing it from the queue
+    func runQueuedRequests(completion: @escaping () -> Void) {
```
Comment (evsaev) [inline Sources/StreamChat/Repositories/OfflineRequestsRepository.swift:40]: This part is a bit confusing. If the goal is to not execute any API requests while the recovery flow, why it's not enough to just add them to `operationQueue` which gets suspended when we enter recovery flow and resumed when recovery is done?

File: Sources/StreamChat/Repositories/OfflineRequestsRepository.swift [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        var pendingActions: [(String, Data)] = []
+
+        let readContext = database.backgroundReadOnlyContext
+        readContext.performAndWait {
```
[the remark this answers is not in the document]
Comment (nuno-vieira) [reply Sources/StreamChat/Repositories/OfflineRequestsRepository.swift:44]: Yes, good point, we need to make sure we don't run this from the MainThread. But sync this is part of the Sync Operations, I don't think that will be the case?

File: Sources/StreamChat/Repositories/OfflineRequestsRepository.swift [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+            pendingActions = QueuedRequestDTO.loadAllPendingRequests(context: readContext).map {
+                ($0.id, $0.endpoint)
+            }
```
Comment (nuno-vieira) [inline Sources/StreamChat/Repositories/OfflineRequestsRepository.swift:47]: Shouldn't we read these pending requests from the retry queue as well? What happens if meanwhile, we add an offline request to the database while we are reading from it? Won't there be any concurrency issues?

**Review by nuno-vieira — COMMENTED** [review]
Looks really good overall 👍  Just added some comments and questions 🙂

File: Sources/StreamChat/Repositories/OfflineRequestsRepository.swift [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+    ) {
+        // TODO: Temporary String check. We only process the response for sent messages
+        guard endpoint.path.hasSuffix("/message") else {
+            completion()
+            return
+        }
+        completion()
+    }
+
+    func queueOfflineRequest(endpoint: DataEndpoint, completion: (() -> Void)? = nil) {
```
Comment (martinmitrevski) [inline Sources/StreamChat/Repositories/OfflineRequestsRepository.swift:111]: Does it makes sense to provide more info in the completion block? At the moment the caller doesn't know which of the different cases called it.

File: Sources/StreamChat/APIClient/APIClient.swift [hunk 7]
```diff
@@ -220,18 +237,14 @@ class APIClient {
...
     }
 
-    private func shouldRetry(_ error: Error, operation: AsyncOperation) -> Bool {
+    private func isConnectionError(_ error: Error) -> Bool {
         // We only retry transient errors like connectivity stuff or HTTP 5xx errors
         guard ClientError.isEphemeral(error: error) else {
             return false
         }
 
         let offlineErrorCodes: Set<Int> = [NSURLErrorDataNotAllowed, NSURLErrorNotConnectedToInternet]
```
Comment (martinmitrevski) [inline Sources/StreamChat/APIClient/APIClient.swift:246]: Are these enough for offline error codes? I would suggest going through the URLError.Codes to possibly include others too (e.g. `networkConnectionLost`, `internationalRoamingOff` etc). 
Also, I think it's better to use the more Swifty syntax here (`URLError.Code.notConnectedToInternet`).

**Review by martinmitrevski — COMMENTED** [review]
Looks great in general! Added some comments and @nuno-vieira and @evsaev also have some valid remarks worth checking.

File: Sources/StreamChat/Repositories/OfflineRequestsRepository.swift (hunk 3 above)
Comment (polqf) [reply Sources/StreamChat/Repositories/OfflineRequestsRepository.swift:40]: Discussed offline

File: Sources/StreamChat/APIClient/APIClient.swift [hunk 8]
```diff
@@ -118,8 +122,21 @@ class APIClient {
...
                     done(.retry)
                     operation.resetRetries()
-                case let .failure(error) where self?.shouldRetry(error, operation: operation) == true:
-                    done(.retry)
+                case let .failure(error) where self?.isConnectionError(error) == true:
+                    // Do not retry unless its a connection problem and we still have retries left
+                    if operation.canRetry {
+                        done(.retry)
+                        return
+                    }
```
[the remark this answers is not in the document]
Comment (polqf) [reply Sources/StreamChat/APIClient/APIClient.swift:130]: Done in further PR

File: Sources/StreamChat/APIClient/APIClient.swift [hunk 9]
```diff
@@ -118,8 +122,21 @@ class APIClient {
                     // Retry request. Expired token has been refreshed
                     done(.retry)
                     operation.resetRetries()
```
[the remark this answers is not in the document]
Comment (polqf) [reply Sources/StreamChat/APIClient/APIClient.swift:124]: Discussed offline, and fixed in following PRs

File: Sources/StreamChat/APIClient/Endpoints/Endpoint.swift [hunk 10]
```diff
@@ -27,9 +27,50 @@ struct Endpoint<ResponseType: Decodable> {
...
+    }
+
+    init(from decoder: Decoder) throws {
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+        path = try container.decode(String.self, forKey: .path)
+        method = try container.decode(EndpointMethod.self, forKey: .method)
+        queryItems = try container.decodeIfPresent(Data.self, forKey: .queryItems)
+        requiresConnectionId = try container.decode(Bool.self, forKey: .requiresConnectionId)
+        requiresToken = try container.decode(Bool.self, forKey: .requiresToken)
+        body = try container.decodeIfPresent(Data.self, forKey: .body)
```
[the remark this answers is not in the document]
Comment (polqf) [reply Sources/StreamChat/APIClient/Endpoints/Endpoint.swift:49]: This allows for better debugging

File: Sources/StreamChat/Repositories/SyncRepository_Tests.swift [hunk 11]
```diff
@@ -408,4 +419,62 @@ final class SyncRepository_Tests: XCTestCase {
...
+            path: "",
+            method: .post,
+            queryItems: nil,
+            requiresConnectionId: true,
+            requiresToken: true,
+            body: nil
+        )
+        repository.queueOfflineRequest(endpoint: endpoint)
+
+        XCTAssertNotCall("queueOfflineRequest(endpoint:completion:)", on: offlineRequestsRepository)
```
[the remark this answers is not in the document]
Comment (polqf) [reply Sources/StreamChat/Repositories/SyncRepository_Tests.swift:450]: As discussed in the tech meeting, completely agree with needed some standarization. As of now, let's keep it, and reevaluate once there's a final standard in place

File: Sources/StreamChat/APIClient/APIClient.swift (hunk 7 above)
Comment (polqf) [reply Sources/StreamChat/APIClient/APIClient.swift:246]: Actually, this is "improved" in a following PR where only `ClientError.isEphemeral` is used. ATM we are checking codes differently in 3 parts of the LLC...

File: Sources/StreamChat/Repositories/OfflineRequestsRepository.swift (hunk 5 above)
Comment (polqf) [reply Sources/StreamChat/Repositories/OfflineRequestsRepository.swift:47]: You cannot actually do that as the contexts have their own queue. Even if I move it to the retry queue, they will be then moved to the context's queue, leaving the end result as it is right now

**Review by martinmitrevski — APPROVED** [review]

File: Sources/StreamChat/Repositories/OfflineRequestsRepository.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (polqf) [reply Sources/StreamChat/Repositories/OfflineRequestsRepository.swift:44]: Yep, exactly what Nuno mentioned. But I will add a note to improve it in the following PRs

**Review by nuno-vieira — APPROVED** [review]
LGTM! ✅

**Pull request merged**